File: leetcode/company-wise/microsoft/0380_insert_delete_getrandom.py
# LC 380. Insert Delete GetRandom O(1) | Medium | Microsoft
import random

class RandomizedSet:
    def __init__(self): self.lst = []; self.idx = {}
    def insert(self, val):
        if val in self.idx: return False
        self.idx[val] = len(self.lst); self.lst.append(val); return True
    def remove(self, val):
        if val not in self.idx: return False
        last = self.lst[-1]; i = self.idx[val]
        self.lst[i] = last; self.idx[last] = i
        self.lst.pop(); del self.idx[val]; return True
    def getRandom(self): return random.choice(self.lst)

File: leetcode/company-wise/microsoft/0380_insert_delete_getrandom.cpp
// LC 380. Insert Delete GetRandom O(1) | Medium | Microsoft
#include <vector>
#include <unordered_map>
using namespace std;

class RandomizedSet {
    vector<int> v; unordered_map<int,int> mp;
public:
    bool insert(int val) { if (mp.count(val)) return false; mp[val]=v.size(); v.push_back(val); return true; }
    bool remove(int val) {
        if (!mp.count(val)) return false;
        int i = mp[val], last = v.back();
        v[i] = last; mp[last] = i; v.pop_back(); mp.erase(val); return true;
    }
    int getRandom() { return v[rand()%v.size()]; }
};
